File: binding/src/libqstringmatcher.cpp
#include <QChar>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QStringMatcher>
#include <qstringmatcher.h>
#include "libqstringmatcher.h"
#include "libqstringmatcher.hxx"

QStringMatcher* QStringMatcher_new() {
	 return new QStringMatcher();
}

QStringMatcher* QStringMatcher_new2(const libqt_string pattern) {
	 return new QStringMatcher(QString::fromUtf8(pattern.data, pattern.len));
}

QStringMatcher* QStringMatcher_new3(const QChar* uc, ptrdiff_t lenVal) {
	 return new QStringMatcher(uc, lenVal);
}

QStringMatcher* QStringMatcher_new4(const QStringMatcher* other) {
	 return new QStringMatcher(*other);
}

QStringMatcher* QStringMatcher_new5(const libqt_string pattern, int cs) {
	 return new QStringMatcher(QString::fromUtf8(pattern.data, pattern.len), static_cast<Qt::CaseSensitivity>(cs));
}

QStringMatcher* QStringMatcher_new6(const QChar* uc, ptrdiff_t lenVal, int cs) {
	 return new QStringMatcher(uc, lenVal, static_cast<Qt::CaseSensitivity>(cs));
}

void QStringMatcher_OperatorAssign(QStringMatcher* self, const QStringMatcher* other) {
	self->operator=(*other);
}

void QStringMatcher_SetPattern(QStringMatcher* self, const libqt_string pattern) {
	self->setPattern(QString::fromUtf8(pattern.data, pattern.len));
}

void QStringMatcher_SetCaseSensitivity(QStringMatcher* self, int cs) {
	self->setCaseSensitivity(static_cast<Qt::CaseSensitivity>(cs));
}

ptrdiff_t QStringMatcher_IndexIn(const QStringMatcher* self, const libqt_string str) {
	return self->indexIn(QString::fromUtf8(str.data, str.len));
}

ptrdiff_t QStringMatcher_IndexIn2(const QStringMatcher* self, const QChar* str, ptrdiff_t length) {
	return self->indexIn(str, length);
}

libqt_string QStringMatcher_Pattern(const QStringMatcher* self) {
	QString _ret = self->pattern();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QStringMatcher_CaseSensitivity(const QStringMatcher* self) {
	return self->caseSensitivity();
}

ptrdiff_t QStringMatcher_IndexIn22(const QStringMatcher* self, const libqt_string str, ptrdiff_t from) {
	return self->indexIn(QString::fromUtf8(str.data, str.len), from);
}

ptrdiff_t QStringMatcher_IndexIn32(const QStringMatcher* self, const QChar* str, ptrdiff_t length, ptrdiff_t from) {
	return self->indexIn(str, length, from);
}

void QStringMatcher_Delete(QStringMatcher* self) {
    delete self;
}

File: binding/include/libqstringmatcher.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qstringmatcher.html
pub const qstringmatcher = struct {

    /// New constructs a new QStringMatcher object.
    pub fn New() QtC.QStringMatcher {
        return qtc.QStringMatcher_new();
    }


    /// New2 constructs a new QStringMatcher object.
    pub fn New2(pattern: []const u8) QtC.QStringMatcher {
        const pattern_str = qtc.libqt_string{
    .len = pattern.len,
    .data = pattern.ptr,
};
return qtc.QStringMatcher_new2(pattern_str);
    }


    /// New3 constructs a new QStringMatcher object.
    pub fn New3(uc: ?*anyopaque, lenVal: i64) QtC.QStringMatcher {
        return qtc.QStringMatcher_new3(@ptrCast(uc), lenVal);
    }


    /// New4 constructs a new QStringMatcher object.
    pub fn New4(pattern: []const u8) QtC.QStringMatcher {
        return qtc.QStringMatcher_new4(@ptrCast(pattern));
    }


    /// New5 constructs a new QStringMatcher object.
    pub fn New5(other: ?*anyopaque) QtC.QStringMatcher {
        return qtc.QStringMatcher_new5(@ptrCast(other));
    }


    /// New6 constructs a new QStringMatcher object.
    pub fn New6(pattern: []const u8, cs: i32) QtC.QStringMatcher {
        const pattern_str = qtc.libqt_string{
    .len = pattern.len,
    .data = pattern.ptr,
};
return qtc.QStringMatcher_new6(pattern_str, @intCast(cs));
    }


    /// New7 constructs a new QStringMatcher object.
    pub fn New7(uc: ?*anyopaque, lenVal: i64, cs: i32) QtC.QStringMatcher {
        return qtc.QStringMatcher_new7(@ptrCast(uc), lenVal, @intCast(cs));
    }


    /// New8 constructs a new QStringMatcher object.
    pub fn New8(pattern: []const u8, cs: i32) QtC.QStringMatcher {
        return qtc.QStringMatcher_new8(@ptrCast(pattern), @intCast(cs));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QStringMatcher_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn SetPattern(self: ?*anyopaque, pattern: []const u8) void {
        const pattern_str = qtc.libqt_string{
    .len = pattern.len,
    .data = pattern.ptr,
};
qtc.QStringMatcher_SetPattern(@ptrCast(self), pattern_str);
    }

    pub fn SetCaseSensitivity(self: ?*anyopaque, cs: i32) void {
        qtc.QStringMatcher_SetCaseSensitivity(@ptrCast(self), @intCast(cs));
    }

    pub fn IndexIn(self: ?*anyopaque, str: []const u8) i64 {
        const str_str = qtc.libqt_string{
    .len = str.len,
    .data = str.ptr,
};
return qtc.QStringMatcher_IndexIn(@ptrCast(self), str_str);
    }

    pub fn IndexIn2(self: ?*anyopaque, str: ?*anyopaque, length: i64) i64 {
        return qtc.QStringMatcher_IndexIn2(@ptrCast(self), @ptrCast(str), length);
    }

    pub fn IndexIn3(self: ?*anyopaque, str: []const u8) i64 {
        return qtc.QStringMatcher_IndexIn3(@ptrCast(self), @ptrCast(str));
    }

    pub fn Pattern(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QStringMatcher_Pattern(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstringmatcher.Pattern: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn PatternView(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QStringMatcher_PatternView(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qstringmatcher.PatternView: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn CaseSensitivity(self: ?*anyopaque, ) i32 {
        return qtc.QStringMatcher_CaseSensitivity(@ptrCast(self));
    }

    pub fn IndexIn22(self: ?*anyopaque, str: []const u8, from: i64) i64 {
        const str_str = qtc.libqt_string{
    .len = str.len,
    .data = str.ptr,
};
return qtc.QStringMatcher_IndexIn22(@ptrCast(self), str_str, from);
    }

    pub fn IndexIn32(self: ?*anyopaque, str: ?*anyopaque, length: i64, from: i64) i64 {
        return qtc.QStringMatcher_IndexIn32(@ptrCast(self), @ptrCast(str), length, from);
    }

    pub fn IndexIn23(self: ?*anyopaque, str: []const u8, from: i64) i64 {
        return qtc.QStringMatcher_IndexIn23(@ptrCast(self), @ptrCast(str), from);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QStringMatcher_Delete(@ptrCast(self));
    }
};
